README.md
backup Cookbook
===============
Chef cookbook to wrap backup rock, available at:
https://github.com/danfruehauf/backup

The cookbook provides a few simple templates and providers to easily deploy
backup rock on Chef environments.

Requirements
------------
That's the beauty, the cookbook doesn't have any major dependencies, most
likely if you have a linux system with /bin/bash and tar, the cookbook should
run.

Attributes
----------

#### backup::default
<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['backup']['download_url']</tt></td>
    <td>String</td>
    <td>URL to download backup rock from - should be a .tar.gz package</td>
    <td><tt>https://github.com/danfruehauf/backup/tarball/master/</tt></td>
  </tr>
  <tr>
    <td><tt>['backup']['base_dir']</tt></td>
    <td>String</td>
    <td>Base directory for backups</td>
    <td><tt>/home/backup</tt></td>
  </tr>
  <tr>
    <td><tt>['backup']['backup_dir']</tt></td>
    <td>String</td>
    <td>Directory to use for backups</td>
    <td><tt>/home/backup/backup</tt></td>
  </tr>
  <tr>
    <td><tt>['backup']['models_dir']</tt></td>
    <td>String</td>
    <td>Directory to store models generated by this cookbook</td>
    <td><tt>/home/backup/models</tt></td>
  </tr>
  <tr>
    <td><tt>['backup']['bin_dir']</tt></td>
    <td>String</td>
    <td>Directory to deploy the backup rock to</td>
    <td><tt>/home/backup/bin</tt></td>
  </tr>
  <tr>
    <td><tt>['backup']['log_dir']</tt></td>
    <td>String</td>
    <td>Directory for log files</td>
    <td><tt>/var/log/backup</tt></td>
  </tr>
  <tr>
    <td><tt>['backup']['username']</tt></td>
    <td>String</td>
    <td>Username to create for backups</td>
    <td><tt>backup</tt></td>
  </tr>
  <tr>
    <td><tt>['backup']['group']</tt></td>
    <td>String</td>
    <td>Group to create for backups</td>
    <td><tt>backup</tt></td>
  </tr>
  <tr>
    <td><tt>['backup']['backups_to_keep']</tt></td>
    <td>Integer</td>
    <td>How many backups to keep</td>
    <td><tt>3</tt></td>
  </tr>
</table>

Usage
-----
#### backup::default
In your recipes, you should:
 * Include the backup recipe
 * Use the `backup` provider to define your backups

For example:
```
# Include the recipe
include_recipe "backup::default"

# Define files and directories to be backed up
files = []
files.push("/etc/cake")
files.push("/etc/chocolate_cake")

# Define databases to be backed up
databases = []

# A MySQL database
databases.push({
  'type'     => 'mysql',
  'host'     => database_host,
  'port'     => database_port,
  'name'     => database_name,
  'username' => database_username,
  'password' => database_password
})

# A PostgreSQL database
databases.push({
  'type'     => 'pgsql',
  'host'     => database_host,
  'port'     => database_port,
  'name'     => database_name,
  'username' => database_username,
  'password' => database_password
})

# Define a backup model for an application called 'cake' backing up the files
# and database we defined
backup 'cake' do
  databases databases
  files     files
end

```

Creating New Templates
----------------------
Alternatively, if you'd like to wrap that cookbook without 'touching' it, you
can use templates from your own cookbook. In the example below we'll use a
template model `my_template` from the cookbook `my_cookbook`. The template
should be in `my_cookbook/templates/default/my_template.sh.erb`.

Then you can use the backup template from your cookbook:
```
backup 'custom' do
  cookbook  'my_cookbook'
  template  'my_template'
  databases databases
  files     files
end
```

Contributing
------------
1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write you change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
Authors:
 * Dan Fruehauf <malkodan@gmail.com>


CHANGELOG.md
# CHANGELOG for backup

This file is used to list changes made in each version of backup.

## 0.1.0:

* Initial release of backup

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


recipes/default.rb
#
# Cookbook Name:: backup
# Recipe:: default
#
# Copyright (C) 2013 Dan Fruehauf <malkodan@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#

# Require some pretty basic packages
%w{ rsync tar gzip openssh-clients }.each do |package|
  package package
end

# Create backup group
group node[:backup][:group] do
end

# Create backup user
user node[:backup][:username] do
  home  node[:backup][:base_dir]
  gid   node[:backup][:group]
end

# Create all directories with right permissions
[
  node[:backup][:base_dir],
  node[:backup][:backup_dir],
  node[:backup][:models_dir],
  node[:backup][:bin_dir],
  node[:backup][:log_dir]
].each do |directory|
  directory directory do
    mode 0775
    owner node[:backup][:username]
    group node[:backup][:group]
    recursive true
  end
end

# Download and deploy backup executable and plugins
backup_install_path    = "#{node[:backup][:bin_dir]}/backup.sh"
backup_tar_gz_location = "#{Chef::Config[:file_cache_path]}/backup.tar.gz"

# Download backup rock
remote_file backup_tar_gz_location do
  source   node[:backup][:download_url]
  not_if   { ::File.exists?(backup_install_path) }
  notifies :run, "execute[extract_backup_rock]", :immediately
end

# Extract backup rock
execute "extract_backup_rock" do
  command "/bin/tar --strip-components=1 -C #{node[:backup][:bin_dir]} -xf #{backup_tar_gz_location}"
  creates backup_install_path
  action  :nothing
end



providers/backup.rb
#
# Copyright (C) 2013 Dan Fruehauf <malkodan@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#

def whyrun_supported?
  true
end

# Add
action :add do
  # Parameters
  backup_name     = new_resource.backup_name
  params          = new_resource.params
  cookbook        = new_resource.cookbook || "backup"
  template        = new_resource.template || "default"
  model_file_path = "#{node['backup']['models_dir']}/#{backup_name}.sh"

  Chef::Log.info "Adding backup model '#{backup_name} at #{model_file_path}"

  # Create the template
  template model_file_path do
    source   "model_#{template}.sh.erb"
    cookbook cookbook
    owner    node['backup']['username']
    group    node['backup']['group']
    mode     0644
    variables(
      :backup_name => backup_name,
      :params      => params
    )
  end
  new_resource.updated_by_last_action(true)
end

# Remove
action :remove do
  # Parameters
  model_file_path = "#{node['backup']['models_dir']}/#{backup_name}.sh"

  if ::File.exists?(model_file_path)
    Chef::Log.info "Removing #{new_resource.command_name} from #{node['nagios']['nrpe']['conf_dir']}/nrpe.d/"
    file model_file_path do
      action :delete
    end
    new_resource.updated_by_last_action(true)
  end
end

